crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -254,3 +287,284 @@ impl Dispatch<xdg_wm_base::XdgWmBase, ()> for WaylandClientState {
         }
     }
 }
+
+impl Dispatch<wl_seat::WlSeat, ()> for WaylandClientState {
+    fn event(
+        state: &mut Self,
+        seat: &wl_seat::WlSeat,
+        event: wl_seat::Event,
+        data: &(),
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+    ) {
+        if let wl_seat::Event::Capabilities {
+            capabilities: WEnum::Value(capabilities),
+        } = event
+        {
+            if capabilities.contains(wl_seat::Capability::Keyboard) {
+                seat.get_keyboard(&qh, ());
+            }
+            if capabilities.contains(wl_seat::Capability::Pointer) {
+                seat.get_pointer(&qh, ());
+            }
+        }
+    }
+}
+
+impl Dispatch<wl_keyboard::WlKeyboard, ()> for WaylandClientState {
+    fn event(
+        state: &mut Self,
+        keyboard: &wl_keyboard::WlKeyboard,
+        event: wl_keyboard::Event,
+        data: &(),
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+    ) {
+        if let wl_keyboard::Event::Keymap {
+            format: WEnum::Value(format),
+            fd,
+            size,
+            ..
+        } = event
+        {
+            assert_eq!(
+                format,
+                wl_keyboard::KeymapFormat::XkbV1,
+                "Unsupported keymap format"
+            );
+            let keymap = unsafe {
+                xkb::Keymap::new_from_fd(
+                    &xkb::Context::new(xkb::CONTEXT_NO_FLAGS),
+                    fd,
+                    size as usize,
+                    XKB_KEYMAP_FORMAT_TEXT_V1,
+                    KEYMAP_COMPILE_NO_FLAGS,
+                )
+                .unwrap()
+            }
+            .unwrap();
+            state.keymap_state = Some(xkb::State::new(&keymap));
+            state.keymap = Some(keymap);
+        } else if let wl_keyboard::Event::Enter { surface, .. } = event {
+            for window in &state.windows {
+                if window.1.surface.id() == surface.id() {
+                    state.focused_window = Some(Arc::clone(&window.1));
+                }
+            }
+        } else if let wl_keyboard::Event::Key {
+            key,
+            state: WEnum::Value(key_state),
+            ..
+        } = event
+        {
+            let keymap = state.keymap.as_ref().unwrap();
+            let keymap_state = state.keymap_state.as_ref().unwrap();
+            let key_string = keymap_state.key_get_utf8(Keycode::from(key + 8));
```

Comment:
this constant of `8` - it must be declared somewhere, I assume? Would be good to have a descriptive name that explains why it's there in the first place.